File: tasks/configuration-merge/typescript/solution/config_merge.ts
import * as fs from "node:fs";

type JsonValue = null | boolean | number | string | JsonValue[] | JsonObject;
type JsonObject = { [key: string]: JsonValue };
type Input = { defaults: JsonObject; file: JsonObject; env: JsonObject; cli: JsonObject };

function isObject(value: unknown): value is JsonObject {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function validate(value: unknown): asserts value is Input {
  const keys = ["defaults", "file", "env", "cli"];
  if (!isObject(value)) throw new Error("invalid input");
  if (Object.keys(value).sort().join(",") !== keys.slice().sort().join(",")) throw new Error("invalid layers");
  for (const key of keys) if (!isObject(value[key])) throw new Error("invalid layer");
}

function clone(value: JsonValue): JsonValue {
  if (Array.isArray(value)) return value.map(clone);
  if (isObject(value)) return Object.fromEntries(Object.entries(value).map(([key, item]) => [key, clone(item)]));
  return value;
}

function mergeInto(target: JsonObject, layer: JsonObject, sabotage: string): JsonObject {
  for (const [key, value] of Object.entries(layer)) {
    if (value === null && sabotage !== "ignore-delete") {
      delete target[key];
    } else if (isObject(value) && sabotage !== "shallow-merge") {
      const base = target[key];
      target[key] = mergeInto(isObject(base) ? clone(base) as JsonObject : {}, value, sabotage);
    } else if (Array.isArray(value) && sabotage === "merge-arrays" && Array.isArray(target[key])) {
      target[key] = [...target[key] as JsonValue[], ...clone(value) as JsonValue[]];
    } else {
      target[key] = clone(value);
    }
  }
  return target;
}

function merge(value: unknown): JsonObject {
  validate(value);
  const sabotage = process.env.LAB_SABOTAGE || "";
  const layers = [value.defaults, value.file, value.env, value.cli];
  if (sabotage === "reverse-precedence") layers.reverse();
  return layers.reduce((result, layer) => mergeInto(result, layer, sabotage), {});
}

try {
  const input: unknown = JSON.parse(fs.readFileSync(0, "utf8"));
  process.stdout.write(JSON.stringify(merge(input)) + "\n");
} catch (error) {
  process.stderr.write(String(error instanceof Error ? error.message : error) + "\n");
  process.exitCode = 1;
}

File: tasks/configuration-merge/python/solution/config_merge.py
import copy
import json
import os
import sys


def validate(value):
    keys = {"defaults", "file", "env", "cli"}
    if not isinstance(value, dict) or set(value) != keys:
        raise ValueError("invalid input")
    if any(not isinstance(value[key], dict) for key in keys):
        raise ValueError("invalid layer")


def merge_into(target, layer, sabotage):
    for key, value in layer.items():
        if value is None and sabotage != "ignore-delete":
            target.pop(key, None)
        elif isinstance(value, dict) and sabotage != "shallow-merge":
            base = target.get(key)
            target[key] = merge_into(copy.deepcopy(base) if isinstance(base, dict) else {}, value, sabotage)
        elif isinstance(value, list) and sabotage == "merge-arrays" and isinstance(target.get(key), list):
            target[key] = copy.deepcopy(target[key]) + copy.deepcopy(value)
        else:
            target[key] = copy.deepcopy(value)
    return target


def merge(value):
    validate(value)
    sabotage = os.environ.get("LAB_SABOTAGE", "")
    layers = [value[key] for key in ("defaults", "file", "env", "cli")]
    if sabotage == "reverse-precedence":
        layers.reverse()
    result = {}
    for layer in layers:
        merge_into(result, layer, sabotage)
    return result


try:
    print(json.dumps(merge(json.load(sys.stdin)), separators=(",", ":")))
except Exception as error:
    print(str(error), file=sys.stderr)
    raise SystemExit(1)
